### ISSUE TITLE
Track styling for gene URL query

### ISSUE THREAD

**Issue by kpepper** [GMOD/Apollo#1942]
I want to be able to display a specific gene on an organism with public mode set, using a URL query string. So, I thought I could use something along the lines of:

domain/organism name/jbrowse/index.html?data=organism data folder&loc=gene name

This seems to display the relevant gene as expected but displays the tracks using non-neat html features. I was expecting the same styling as used on the public organism window (neat, in my case). Any ideas how to circumvent this? Specifying a chromosome:from..to format for the _loc_ parameter and no _data_ parameter shows the expected styling. If I don't specify the _data_ parameter when passing a feature id then it fails to find the feature.

Thanks.

**Comment by nathandunn** [comment]
Interesting.   It may not respect the plugins, but not sure how. 

That being said, the "data=" should be redundant with the organism name in the URL and not necessary (its appended internally) and may screw stuff up. 

If this doesn't work (removing "data=" and see below), you can also just create a JBrowse instance that mounts to these same directories, but hopefully this works.  Let us know either way.



There is a link that is supposed to supply the same with the current view:

- <img width="750" alt="screen shot 2018-10-01 at 9 14 46 am" src="https://user-images.githubusercontent.com/751274/46301313-d1a85a80-c55a-11e8-946d-239578966d6c.png">

- <img width="869" alt="screen shot 2018-10-01 at 9 14 51 am" src="https://user-images.githubusercontent.com/751274/46301312-d10fc400-c55a-11e8-9ce3-28d3ccc4c444.png">

My resulting link:
- http://demo.genomearchitect.org/Apollo2/21354576081831563254/jbrowse/index.html?loc=Group11.18:2365822..2390238&tracks=Official%20Gene%20Set%20v3.2,Augustus%20Set%2012,Nurse,Apis%20cerana%20RNA-Seq%20reads

**Comment by kpepper** [comment]
It seems to work sometimes, not quite sure why [without the _data_ parameter] but it looks like I'm getting a javascript exception in the web console:

Browser.js:2022 TypeError: Cannot read property 'query' of undefined TypeError: Cannot read property 'query' of undefined
    at Object.searchNames (https://website/organism_name/jbrowse/src/JBrowse/Browser.js:2539:27)
    at https://website/organism_name/jbrowse/src/JBrowse/Browser.js:2419:15
    at https://website/organism_name/jbrowse/src/JBrowse/Browser.js:2020:29
    at signalListener (https://website/organism_name/jbrowse/src/dojo/Deferred.js:37:21)
    at signalWaiting (https://website/organism_name/jbrowse/src/dojo/Deferred.js:28:4)
    at Deferred.resolve (https://website/organism_name/jbrowse/src/dojo/Deferred.js:192:5)
    at Object.<anonymous> (https://website/organism_name/jbrowse/src/JBrowse/Browser.js:992:30)
    at https://website/organism_name/jbrowse/src/dojo/_base/lang.js:385:55
    at signalListener (https://website/organism_name/jbrowse/src/dojo/Deferred.js:37:21)
    at Deferred.then.promise.then (https://website/organism_name/jbrowse/src/dojo/Deferred.js:258:5) TypeError: Cannot read property 'query' of undefined
    at Object.searchNames (Browser.js:2539)
    at Browser.js:2419
    at Browser.js:2020
    at signalListener (Deferred.js:37)
    at signalWaiting (Deferred.js:28)
    at Deferred.resolve (Deferred.js:192)
    at Object.<anonymous> (Browser.js:992)
    at lang.js:385
    at signalListener (Deferred.js:37)
    at Deferred.then.promise.then (Deferred.js:258)
xhr.js:440 ErrorCtor {message: "Unable to load ../AnnotationEditorService status: 401", response: {…}, status: 401, responseText: "{"message":"You must first login before editing"}", xhr: XMLHttpRequest, …}

Using the public link and then entering the URL outside of a logged in session gives the same error and behaviour. Entering the gene name directly into the search text field when logged in zooms to the correct feature with no errors.

**Comment by cmdcolin** [comment]
There was a user on the JBrowse forum also reporting this for their WebApollo instance where it would navigate to ?loc=genename properly on public jbrowse if not-logged-in as annotator but I think failed when logged-in-as-annotator https://github.com/GMOD/jbrowse/issues/899

**Comment by nathandunn** [comment]
My guess is its trying to use the nameStore. 

Do you have that defined? 

Also, what is your URL? 

Nathan

**Comment by kpepper** [comment]
**Example query URL:**

https://apollo.genedb.org/7383/jbrowse/index.html?loc=REV3&tracks=DNA%2CAnnotations%2Cgene%2CmRNA&highlight=

**My guess is its trying to use the nameStore. Do you have that defined?**

I run generate_names.pl - the following is defined in the trackList.json files:

"names" : {
      "type" : "Hash",
      "url" : "names/"
   },

and every organism has a names folder with json files in.
I'm not aware of any other custom name store settings. The text field search works as expected, logged in or not.

**Comment by nathandunn** [comment]
Your location (?loc=REV3) should be defined as:

<chr>:<start>..<ned>

e.g., 

Chr3:121131..112411

Or is REV3 short for something else? 

This may be a bug in our code, as well.  Are you using 2.1.0? 

Nathan

**Comment by cmdcolin** [comment]
@nathandunn in JBrowse world you can use gene names in the ?loc= string

**Comment by nathandunn** [comment]
Sorry, ignore my previous comment @kpepper    This is a feature request and not a bug.    

Sorry, I misread this.  I'll see what I can do, as it definitely seems doable, but am mistreating the input.  Probably won't be until next week, though.

@cmdcolin  Was this a newer feature (post 1.12?).  It seems like this could have been working for quite a while.

**Comment by nathandunn** [comment]
In the interim @kpepper  You could just up a separate JBrowse instance over the same set of directories.

**Comment by cmdcolin** [comment]
Having feature names in ?loc= is an old feature :)

**Comment by kpepper** [comment]
Thanks for the assistance with this. I had assumed that because you can query on gene name (i.e. REV3 in the case above) in JBrowse it would automatically be carried into Web Apollo.  It would be great if you could add this as it's very useful functionality for us.

**Comment by kpepper** [comment]
Hi, just wondered if this will be in the next release?

**Comment by nathandunn** [comment]
I'll try to take a look today or tomorrow

**Comment by nathandunn** [comment]
@kpepper  This does work for me, but not necessarily for every sequence.  If you have the name store defined, then you should be able to do something like this:

http://localhost:8080/apollo/21/jbrowse/index.html?loc=YBR209W&tracks=Genes

or 

http://localhost:8080/apollo/yeast/jbrowse/index.html?loc=YBR210W&tracks=Genes

(this is using the JBrowse yeast sample data).    Yeast has a processed name store, which is why it works.

**Comment by nathandunn** [comment]
@kpepper  If this works with JBrowse, but not with Apollo, please re-open and send me some sample data so I can get it fixed.  Thanks

**Comment by kpepper** [comment]
It's not working on Apollo for organisms with name stores generated using generate-names.pl.

**Comment by nathandunn** [comment]
I see.    If those work within JBrowse, then maybe the easiest thing (if you are able) would be to send me a portion of the data (a link via email) and I can see what is wrong locally.

Nathan

**Comment by kpepper** [comment]
**@nathandunn** Sent you an email a few days back regarding this issue with some data in.

**Comment by nathandunn** [comment]
Noted @kpepper   thanks.   I would really need to get the organism directory with minimally the refSeq, and gene and mRNA tracks (and the names).   I can take a look at it for 2.2.0 if I get that.  I'll follow up with this in the email as well.

I use something like this to launch JBrowse, but no worries: https://docs.python.org/2/library/simplehttpserver.html

**Comment by nathandunn** [comment]
@kpepper  I just retested with another working "names" field.   It is an odd problem, and seems like it might be a race condition, as I'm not having it occur locally.  

I think the ```name``` part is correct, as it does to lookups properly.   It also works if I pause it via debugging and walk it through slowly (on your site).

It always seems to work locally. 

My suggestion would be top update to master (if you are able to) and see if you continue to have the problem and then I can fit it based directly on your data.

**Comment by trstickland** [comment]
@nathandunn  in case it helps, here's the javascript console output for a 'problem' request
(this example, https://apollo-scratch.genedb.org/Pfalciparum/jbrowse/index.html?loc=PF3D7_0309700.1&tracks=mRNA )

I updated to jbrowse 1.15.4 and it's exactly the same error & console message (aside: still using apollo 2.1.0, can't get 2.2.0 to build though I assume that's our local customization -- still investigating...)

```
loaded WebApollo plugin
main.js:29 plugin: NeatHTMLFeatures
main.js:29 plugin: NeatCanvasFeatures
GenomeView.js:62 >>> view constructor
Browser.js:2022 TypeError: Cannot read property 'query' of undefined TypeError: Cannot read property 'query' of undefined
    at Object.searchNames (http://localhost:8000/10283327342812291871865160466/jbrowse/src/JBrowse/Browser.js:2539:27)
    at http://localhost:8000/10283327342812291871865160466/jbrowse/src/JBrowse/Browser.js:2419:15
    at http://localhost:8000/10283327342812291871865160466/jbrowse/src/JBrowse/Browser.js:2020:29
    at signalListener (http://localhost:8000/10283327342812291871865160466/jbrowse/src/dojo/Deferred.js:37:21)
    at signalWaiting (http://localhost:8000/10283327342812291871865160466/jbrowse/src/dojo/Deferred.js:28:4)
    at Deferred.resolve (http://localhost:8000/10283327342812291871865160466/jbrowse/src/dojo/Deferred.js:192:5)
    at Object.<anonymous> (http://localhost:8000/10283327342812291871865160466/jbrowse/src/JBrowse/Browser.js:992:30)
    at http://localhost:8000/10283327342812291871865160466/jbrowse/src/dojo/_base/lang.js:385:55
    at signalListener (http://localhost:8000/10283327342812291871865160466/jbrowse/src/dojo/Deferred.js:37:21)
    at signalWaiting (http://localhost:8000/10283327342812291871865160466/jbrowse/src/dojo/Deferred.js:28:4) TypeError: Cannot read property 'query' of undefined
    at Object.searchNames (Browser.js:2539)
    at Browser.js:2419
    at Browser.js:2020
    at signalListener (Deferred.js:37)
    at signalWaiting (Deferred.js:28)
    at Deferred.resolve (Deferred.js:192)
    at Object.<anonymous> (Browser.js:992)
    at lang.js:385
    at signalListener (Deferred.js:37)
    at signalWaiting (Deferred.js:28)
(anonymous) @ Browser.js:2022
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ Browser.js:992
(anonymous) @ lang.js:385
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ Browser.js:2370
(anonymous) @ lang.js:385
runFactory @ dojo.js:1175
execModule @ dojo.js:1303
(anonymous) @ dojo.js:1341
guardCheckComplete @ dojo.js:1318
checkComplete @ dojo.js:1336
onLoadCallback @ dojo.js:1504
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
contextRequire @ dojo.js:856
req @ dojo.js:147
(anonymous) @ Browser.js:2345
_milestoneFunction @ Browser.js:1992
createTrackList @ Browser.js:2336
(anonymous) @ Browser.js:980
(anonymous) @ lang.js:385
(anonymous) @ Browser.js:2020
signalListener @ Deferred.js:37
Deferred.then.promise.then @ Deferred.js:258
afterMilestone @ Browser.js:2018
(anonymous) @ Browser.js:979
(anonymous) @ lang.js:385
(anonymous) @ Browser.js:2020
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ Browser.js:544
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:58
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
handleResponse @ xhr.js:81
onLoad @ xhr.js:116
load (async)
addListeners @ xhr.js:144
xhr @ xhr.js:231
(anonymous) @ Browser.js:536
_milestoneFunction @ Browser.js:1992
loadRefSeqs @ Browser.js:520
(anonymous) @ Browser.js:173
signalListener @ Deferred.js:37
Deferred.then.promise.then @ Deferred.js:258
(anonymous) @ Browser.js:170
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ Browser.js:325
notify @ Deferred.js:187
complete @ Deferred.js:168
resolve.callback @ Deferred.js:248
addResult @ DeferredList.js:54
(anonymous) @ DeferredList.js:37
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ Browser.js:358
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ Browser.js:589
finish @ lazyload.js:111
pollWebKit @ lazyload.js:326
setTimeout (async)
pollWebKit @ lazyload.js:335
setTimeout (async)
pollWebKit @ lazyload.js:335
setTimeout (async)
pollWebKit @ lazyload.js:335
setTimeout (async)
pollWebKit @ lazyload.js:335
setTimeout (async)
pollWebKit @ lazyload.js:335
load @ lazyload.js:247
css @ lazyload.js:365
_loadCSS @ Browser.js:589
(anonymous) @ Browser.js:354
forEach @ array.js:247
(anonymous) @ Browser.js:337
(anonymous) @ lang.js:385
runFactory @ dojo.js:1175
execModule @ dojo.js:1303
(anonymous) @ dojo.js:1341
guardCheckComplete @ dojo.js:1318
checkComplete @ dojo.js:1336
onLoadCallback @ dojo.js:1504
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
(anonymous) @ dojo.js:1667
forEach @ dojo.js:116
runDefQ @ dojo.js:1666
onLoadCallback @ dojo.js:1483
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
contextRequire @ dojo.js:856
result @ dojo.js:883
loadRenderer @ renderer.js:81
load @ renderer.js:92
injectPlugin @ dojo.js:1397
(anonymous) @ dojo.js:1207
forEach @ dojo.js:116
resolvePluginLoadQ @ dojo.js:1200
finishExec @ dojo.js:1241
execModule @ dojo.js:1304
(anonymous) @ dojo.js:1341
guardCheckComplete @ dojo.js:1318
checkComplete @ dojo.js:1336
onLoadCallback @ dojo.js:1504
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
injectPlugin @ dojo.js:1407
injectModule @ dojo.js:1477
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
(anonymous) @ dojo.js:1667
forEach @ dojo.js:116
runDefQ @ dojo.js:1666
onLoadCallback @ dojo.js:1483
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
(anonymous) @ dojo.js:1667
forEach @ dojo.js:116
runDefQ @ dojo.js:1666
onLoadCallback @ dojo.js:1483
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
(anonymous) @ dojo.js:1667
forEach @ dojo.js:116
runDefQ @ dojo.js:1666
onLoadCallback @ dojo.js:1483
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
(anonymous) @ dojo.js:1667
forEach @ dojo.js:116
runDefQ @ dojo.js:1666
onLoadCallback @ dojo.js:1483
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
(anonymous) @ dojo.js:1667
forEach @ dojo.js:116
runDefQ @ dojo.js:1666
onLoadCallback @ dojo.js:1483
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
(anonymous) @ dojo.js:1667
forEach @ dojo.js:116
runDefQ @ dojo.js:1666
onLoadCallback @ dojo.js:1483
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
(anonymous) @ dojo.js:1667
forEach @ dojo.js:116
runDefQ @ dojo.js:1666
onLoadCallback @ dojo.js:1483
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
(anonymous) @ dojo.js:1667
forEach @ dojo.js:116
runDefQ @ dojo.js:1666
onLoadCallback @ dojo.js:1483
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
contextRequire @ dojo.js:856
req @ dojo.js:147
(anonymous) @ Browser.js:327
_milestoneFunction @ Browser.js:1992
initPlugins @ Browser.js:267
(anonymous) @ Browser.js:168
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ Browser.js:2074
(anonymous) @ lang.js:385
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ all.js:69
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
(anonymous) @ Deferred.js:70
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ all.js:69
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
(anonymous) @ Deferred.js:70
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ all.js:69
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
(anonymous) @ Deferred.js:70
signalListener @ Deferred.js:37
Deferred.then.promise.then @ Deferred.js:258
signalListener @ Deferred.js:45
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:58
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
(anonymous) @ Deferred.js:70
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:58
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
handleResponse @ xhr.js:81
onLoad @ xhr.js:116
load (async)
addListeners @ xhr.js:144
xhr @ xhr.js:231
load @ JB_json_v1.js:45
(anonymous) @ ConfigManager.js:178
signalListener @ Deferred.js:37
Deferred.then.promise.then @ Deferred.js:258
_loadInclude @ ConfigManager.js:170
(anonymous) @ ConfigManager.js:146
map @ array.js:289
_loadRecur @ ConfigManager.js:144
(anonymous) @ ConfigManager.js:148
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:58
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
(anonymous) @ Deferred.js:70
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:58
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
handleResponse @ xhr.js:81
onLoad @ xhr.js:116
load (async)
addListeners @ xhr.js:144
xhr @ xhr.js:231
load @ JB_json_v1.js:45
(anonymous) @ ConfigManager.js:178
signalListener @ Deferred.js:37
Deferred.then.promise.then @ Deferred.js:258
_loadInclude @ ConfigManager.js:170
(anonymous) @ ConfigManager.js:146
map @ array.js:289
_loadRecur @ ConfigManager.js:144
(anonymous) @ ConfigManager.js:148
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:58
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
(anonymous) @ Deferred.js:70
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:58
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
handleResponse @ xhr.js:81
onLoad @ xhr.js:116
load (async)
addListeners @ xhr.js:144
xhr @ xhr.js:231
load @ JB_json_v1.js:45
(anonymous) @ ConfigManager.js:178
signalListener @ Deferred.js:37
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
signalDeferred @ Deferred.js:81
signalListener @ Deferred.js:52
signalWaiting @ Deferred.js:28
Deferred.resolve @ Deferred.js:192
(anonymous) @ Util.js:262
runFactory @ dojo.js:1175
execModule @ dojo.js:1303
(anonymous) @ dojo.js:1341
guardCheckComplete @ dojo.js:1318
checkComplete @ dojo.js:1336
onLoadCallback @ dojo.js:1504
onLoad @ dojo.js:1757
load (async)
domOn @ dojo.js:1709
req.injectUrl @ dojo.js:1760
injectModule @ dojo.js:1592
forEach @ dojo.js:116
(anonymous) @ dojo.js:793
guardCheckComplete @ dojo.js:1318
injectDependencies @ dojo.js:792
contextRequire @ dojo.js:856
req @ dojo.js:147
loadJS @ Util.js:251
_getConfigAdaptor @ ConfigManager.js:95
_loadInclude @ ConfigManager.js:169
(anonymous) @ ConfigManager.js:146
map @ array.js:289
_loadRecur @ ConfigManager.js:144
_loadIncludes @ ConfigManager.js:163
(anonymous) @ ConfigManager.js:62
getFinalConfig @ ConfigManager.js:78
(anonymous) @ Browser.js:2051
_milestoneFunction @ Browser.js:1992
loadConfig @ Browser.js:2049
(anonymous) @ Browser.js:159
callback @ ready.js:121
onEvent @ ready.js:47
processQ @ domReady.js:49
detectReady @ domReady.js:62

```

**Comment by nathandunn** [comment]
@trstickland  Thanks.    The error is here:

https://github.com/GMOD/jbrowse/blob/1.15.4-release/src/JBrowse/Browser.js#L2611

This is implying that the name store isn't loading properly.  

The query for the name store is undefined.  However, I think this is a threading problem.

<img width="1787" alt="screen shot 2018-12-14 at 9 44 46 am" src="https://user-images.githubusercontent.com/751274/50018771-333a6b80-ff85-11e8-9569-1aa2eae82728.png">


I think I would change the line:

    return this.nameStore.query({ name: loc })

to 

    return thisB.nameStore.query({ name: loc })

as that seems to have the proper nameStore established. 

Again, not sure why that thread doesn't catch up.    Let me know if that works.

**Comment by cmdcolin** [comment]
@trstickland that bug where the nameStore.query was not initialized is actually probably a bug on the jbrowse side of things

It looks like JBrowse version 1.12.3 was ok, and then some re-organizing things released in version 1.12.4 was caused regression.

The cause is due to a deleted line that waited on the afterMilestone('loadNames'... function

If you want to patch your current release, you can change your Browser.js to have this:

```
navigateTo: function(loc) {
    var thisB = this;
    this.afterMilestone( 'initView', function() {
        // lastly, try to search our feature names for it
        this.afterMilestone('loadNames', function() {
            thisB.searchNames( loc )
               .then( function( found ) {
                          if( found )                                                                                                                                                     return;

                          // if it's a foo:123..456 location, go there
                          var location = typeof loc == 'string' ? Util.parseLocString( loc ) :  loc;
                          // only call navigateToLocation() directly if location has start and end, otherwise try and fill in start/end from 'location' cookie
                          if( location && ("start" in location) && ("end" in location)) {
                              thisB.navigateToLocation( location );
                              return;
                          }
                          // otherwise, if it's just a word (or a location with only a ref property), try to figure out what it is
                          else {
                              if( typeof loc != 'string')
                                  loc = loc.ref;

                              // is it just the name of one of our ref seqs?
                              var ref = thisB.findReferenceSequence( loc );
                              if( ref ) {
                                  thisB.navigateToLocation( { ref: ref.name } );
                                  return;
                              }
                          }

                          new InfoDialog(
                              {
                                  title: 'Not found',                                                                                                                                         content: 'Not found: <span class="locString">'+loc+'</span>',
                                  className: 'notfound-dialog'
                              }).show();
                      });
        })
    });
},
```

Hope that helps

**Comment by nathandunn** [comment]
@cmdcolin  do you think that will come about in 1.16 or a 1.15.5 (should I create a JBrowse issue?)

**Comment by cmdcolin** [comment]
Sure. If you would like specific backport releases for example 1.15.5 we can do that too

**Comment by trstickland** [comment]
Much appreciated, will try this ASAP and report back.

**Comment by trstickland** [comment]
Thanks @cmdcolin that appears to work.   Currently I just have it on a development VM (internal access only I'm afraid) and have only superficially tested in manually with 4 or 5 links, but it looks good 👍

**Comment by nathandunn** [comment]
FYI, I merged master to release with 1.16.1 so no need to backport @cmdcolin  Thanks for the fix!
